pxtlib/tilemap.ts

```diff
@@ -356,6 +376,54 @@ namespace pxt {
             for (const id of Object.keys(allJRes)) this.state.takenNames[id] = true;
         }
 
+        loadJres(jres: Map<JRes>, skipDuplicates = false) {
+            jres = inflateJRes(jres)
+
+            const tiles = readTiles(jres, "this", true);
+
+            // If we are loading JRES into an existing project (i.e. in multipart tutorials)
+            // we need to correct the tile ids because the user may have created new tiles
+            // and taken some of the ids that were used by the tutorial author
+            let tileMapping: Map<string> = {};
+
+            if (tiles) {
+                for (const tileset of tiles.tileSets) {
+                    for (const tile of tileset.tiles) {
+                        if (skipDuplicates) {
+                            const existing = this.resolveTileByBitmap(tile.bitmap);
+                            if (existing) {
+                                tileMapping[tile.id] = existing.id;
+                                continue;
+                            }
+                        }
+
+                        const newTile = this.createNewTile(tile.bitmap, tile.id);
+
+                        if (newTile.id !== tile.id) {
+                            tileMapping[tile.id] = newTile.id;
+                        }
+                    }
+                }
+            }
+
+            const maps = getTilemaps(jres)
+                .map(tm => ({
+                    id: tm.id,
+                    data: decodeTilemap(tm, id => {
+                        if (tileMapping[id]) {
+                            id = tileMapping[id];
+                        }
+
+                        return this.resolveTile(id)
+                    })
+                }));
+
+            this.state.projectTilemaps.push(...maps);
+
+            for (const id of Object.keys(jres)) this.state.takenNames[id] = true;
```

Comment:
is it possible for these tile ids to have changed from createNewTile?

Reply:
good catch!